### PR TITLE
refactoring Action.nonDetExc into case class

### LINKED ISSUE #105: Action.nonDetExc should be refactored from a tuple into a small class
Action.nonDetExc is currently a 3-tuple, with the last element being a pair again. This makes the code hard to read. A simple class should store the data of the three-tuple with meaningful field names instead of _x.

### PR DESCRIPTION
**Pull request by jyoo980**
Added a case class to represent DetExc, included a companion object for easy construction + future extensibility.

Should resolve #105

### CODE REVIEW

File: src/main/scala/modbat/dsl/Action.scala [hunk 1]
```diff
@@ -3,14 +3,11 @@ package modbat.dsl
 import java.lang.reflect.Method
 import scala.collection.mutable.ListBuffer
 import scala.util.matching.Regex
-import modbat.log.Log
-import modbat.mbt.Main
```
Comment (jyoo980) [inline src/main/scala/modbat/dsl/Action.scala:7]: These imports were unused.

File: src/main/scala/modbat/dsl/DetExc.scala [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package modbat.dsl
+
+import scala.util.matching.Regex
+
+final case class DetExc(excName: Regex, target: State, fullName: String, line: Int)
```
Comment (jyoo980) [inline src/main/scala/modbat/dsl/DetExc.scala:5]: Did the best I can with respect to naming by trying to extract context from the source.

Should be trivial to rename if needed.

**Comment by cyrille-artho** [comment]
Hi James,
Thank you very much for your contributions. It looks like the CI builds from Travis are stuck, and I'm not sure why. I'll have to run the build manually on my computer to see if it works (I can't see any way to start or restart the Travis build).
This will take a bit of time but I'll look into this as soon as I can.

**Comment by cyrille-artho** [comment]
Unfortunately, I get a build error on my local machine:
```
./gradlew clean test
```
...
```
> Task :compileScala
Pruning sources from previous analysis, due to incompatible CompileSetup.
/private/tmp/modbat/src/main/scala/modbat/dsl/Action.scala:42: ambiguous reference to overloaded definition,
both method apply in object DetExc of type (excName: scala.util.matching.Regex, target: modbat.dsl.State, fullName: String, line: Int)modbat.dsl.DetExc
and  method apply in object DetExc of type (excName: scala.util.matching.Regex, target: modbat.dsl.State, fullName: String, line: Int)modbat.dsl.DetExc
match argument types (scala.util.matching.Regex,modbat.dsl.State,String,Int)
          DetExc(new Regex(name), new State(state), fullName.value, line.value)
          ^
/private/tmp/modbat/src/main/scala/modbat/dsl/DetExc.scala:5: method apply is defined twice
  conflicting symbols both originated in file '/private/tmp/modbat/src/main/scala/modbat/dsl/DetExc.scala'
final case class DetExc(excName: Regex, target: State, fullName: String, line: Int)
                 ^
two errors found

> Task :compileScala FAILED

FAILURE: Build failed with an exception.
```

Can you please check what causes this?

**Comment by jyoo980** [comment]
> Unfortunately, I get a build error on my local machine:
> 
> ```
> ./gradlew clean test
> ```
> 
> ...
> 
> ```
> > Task :compileScala
> Pruning sources from previous analysis, due to incompatible CompileSetup.
> /private/tmp/modbat/src/main/scala/modbat/dsl/Action.scala:42: ambiguous reference to overloaded definition,
> both method apply in object DetExc of type (excName: scala.util.matching.Regex, target: modbat.dsl.State, fullName: String, line: Int)modbat.dsl.DetExc
> and  method apply in object DetExc of type (excName: scala.util.matching.Regex, target: modbat.dsl.State, fullName: String, line: Int)modbat.dsl.DetExc
> match argument types (scala.util.matching.Regex,modbat.dsl.State,String,Int)
>           DetExc(new Regex(name), new State(state), fullName.value, line.value)
>           ^
> /private/tmp/modbat/src/main/scala/modbat/dsl/DetExc.scala:5: method apply is defined twice
>   conflicting symbols both originated in file '/private/tmp/modbat/src/main/scala/modbat/dsl/DetExc.scala'
> final case class DetExc(excName: Regex, target: State, fullName: String, line: Int)
>                  ^
> two errors found
> 
> > Task :compileScala FAILED
> 
> FAILURE: Build failed with an exception.
> ```
> 
> Can you please check what causes this?

Looks like it was the duplicate `apply` method in the companion object. This should be fixed now.

File: src/main/scala/modbat/dsl/DetExc.scala (hunk 2 above)
Comment (cyrille-artho) [reply src/main/scala/modbat/dsl/DetExc.scala:5]: The previous renaming `NonDetExc` was the appropriate one; please name this new class `NonDetExc` again.

**Review by cyrille-artho — COMMENTED** [review]
Sorry, I was on the wrong branch when I tried your changes (locally). I'll try again to double-check.
Please rename `DetExc` (back to) `NonDetExc`.

**Comment by cyrille-artho** [comment]
Looks good! There is a merge conflict in `src/main/scala/modbat/dsl/Action.scala` from your other PR (which I have already accepted). Could you please update this PR for it? It's probably best to *rebase* (not merge) this PR based on the earlier change you have made in the other one.

**Comment by cyrille-artho** [comment]
Great, I was able to merge it now, many thanks!

**Pull request merged**